cpp/src/lookup/lookup_src_dst_impl.cuh

```diff
@@ -0,0 +1,1076 @@
+  raft::handle_t const& handle,
+  raft::device_span<edge_id_t const> edge_ids_to_lookup,
+  raft::device_span<edge_type_t const> edge_types_to_lookup,
+  bool multi_gpu) const
+{
+  return pimpl->src_dst_from_edge_id_and_type(
+    handle, edge_ids_to_lookup, edge_types_to_lookup, multi_gpu);
+}
+
+namespace detail {
+
+template <typename GraphViewType,
+          typename EdgeIdInputWrapper,
+          typename EdgeTypeInputWrapper,
+          typename EdgeTypeAndIdToSrcDstLookupContainerType>
+EdgeTypeAndIdToSrcDstLookupContainerType build_edge_id_and_type_to_src_dst_lookup_map(
+  raft::handle_t const& handle,
+  GraphViewType const& graph_view,
+  EdgeIdInputWrapper edge_id_view,
+  EdgeTypeInputWrapper edge_type_view)
+{
+  static_assert(!std::is_same_v<typename EdgeIdInputWrapper::value_type, thrust::nullopt_t>,
+                "Can not create edge id lookup table without edge ids");
+
+  using vertex_t    = typename GraphViewType::vertex_type;
+  using edge_t      = typename GraphViewType::edge_type;
+  using edge_type_t = typename EdgeTypeInputWrapper::value_type;
+  using edge_id_t   = typename EdgeIdInputWrapper::value_type;
+  using value_t     = typename EdgeTypeAndIdToSrcDstLookupContainerType::value_type;
+
+  constexpr bool multi_gpu = GraphViewType::is_multi_gpu;
+  static_assert(std::is_integral_v<edge_type_t>);
+  static_assert(std::is_integral_v<edge_id_t>);
+  static_assert(std::is_same_v<value_t, thrust::tuple<vertex_t, vertex_t>>);
+
+  static_assert(
+    std::is_same_v<typename EdgeTypeAndIdToSrcDstLookupContainerType::edge_type_type, edge_type_t>,
+    "edge_type_t must match with EdgeTypeAndIdToSrcDstLookupContainerType::edge_type_type");
+
+  static_assert(
+    std::is_same_v<typename EdgeTypeAndIdToSrcDstLookupContainerType::edge_id_type, edge_id_t>,
+    "edge_id_t must match with typename EdgeTypeAndIdToSrcDstLookupContainerType::edge_id_type");
+
+  std::vector<edge_type_t> h_types_to_this_gpu{};
+  std::vector<edge_t> h_freq_of_types_to_this_gpu{};
+  std::unordered_map<edge_type_t, edge_t> edge_type_to_count_map{};
+
+  for (size_t local_ep_idx = 0; local_ep_idx < graph_view.number_of_local_edge_partitions();
+       ++local_ep_idx) {
+    //
+    //  Copy edge ids and types
+    //
+
+    auto number_of_edges_partition =
+      graph_view.local_edge_partition_view(local_ep_idx).number_of_edges();
+    auto number_of_active_edges_partition = number_of_edges_partition;
+
+    if (graph_view.has_edge_mask()) {
+      number_of_active_edges_partition =
+        detail::count_set_bits(handle,
+                               (*(graph_view.edge_mask_view())).value_firsts()[local_ep_idx],
+                               number_of_edges_partition);
+    }
+
+    [[maybe_unused]] auto edgelist_ids = std::make_optional<rmm::device_uvector<edge_t>>(
+      number_of_active_edges_partition, handle.get_stream());
+
+    auto edgelist_types = std::make_optional<rmm::device_uvector<edge_type_t>>(
```

Comment:
Is there a reason to make this optional?  It's always defined, and you only ever use it.  Seems like just defining it as an `rmm::device_uvector<edge_type_t>` would be sufficient.

Reply:
Changed.